state/cluster/badger/mutator.go

```diff
@@ -36,202 +37,286 @@ func NewMutableState(state *State, tracer module.Tracer, headers storage.Headers
+	refBlock, err := m.headers.ByBlockID(payload.ReferenceBlockID)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return state.NewUnverifiableExtensionError("cluster block references unknown reference block (id=%x)", payload.ReferenceBlockID)
 		}
+		return fmt.Errorf("could not check reference block: %w", err)
+	}
 
-		checkAnsSpan.End()
-		checkTxsSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendCheckTransactionsValid)
-		defer checkTxsSpan.End()
-
-		// a valid collection must reference a valid reference block
-		// NOTE: it is valid for a collection to be expired at this point,
-		// otherwise we would compromise liveness of the cluster.
-		refBlock, err := m.headers.ByBlockID(payload.ReferenceBlockID)
+	// 2 - the reference block must be finalized
+	if refBlock.Height > ctx.finalizedConsensusHeight {
+		// a reference block which is above the finalized boundary can't be verified yet
+		return state.NewUnverifiableExtensionError("reference block is above finalized boundary (%d>%d)", refBlock.Height, ctx.finalizedConsensusHeight)
+	} else {
+		storedBlockIDForHeight, err := m.headers.BlockIDByHeight(refBlock.Height)
 		if err != nil {
-			if errors.Is(err, storage.ErrNotFound) {
-				return state.NewUnverifiableExtensionError("cluster block references unknown reference block (id=%x)", payload.ReferenceBlockID)
-			}
-			return fmt.Errorf("could not check reference block: %w", err)
+			return irrecoverable.NewExceptionf("could not look up block ID for finalized height: %w", err)
 		}
-
-		// no validation of transactions is necessary for empty collections
-		if payload.Collection.Len() == 0 {
-			return nil
+		// a reference block with height at or below the finalized boundary must have been finalized
+		if storedBlockIDForHeight != payload.ReferenceBlockID {
+			return state.NewInvalidExtensionErrorf("cluster block references orphaned reference block (id=%x, height=%d), the block finalized at this height is %x",
+				payload.ReferenceBlockID, refBlock.Height, storedBlockIDForHeight)
 		}
+	}
 
-		// check that all transactions within the collection are valid
-		// keep track of the min/max reference blocks - the collection must be non-empty
-		// at this point so these are guaranteed to be set correctly
-		minRefID := flow.ZeroID
-		minRefHeight := uint64(math.MaxUint64)
-		maxRefHeight := uint64(0)
-		for _, flowTx := range payload.Collection.Transactions {
-			refBlock, err := m.headers.ByBlockID(flowTx.ReferenceBlockID)
-			if errors.Is(err, storage.ErrNotFound) {
-				// unknown reference blocks are invalid
-				return state.NewUnverifiableExtensionError("collection contains tx (tx_id=%x) with unknown reference block (block_id=%x): %w", flowTx.ID(), flowTx.ReferenceBlockID, err)
-			}
-			if err != nil {
-				return fmt.Errorf("could not check reference block (id=%x): %w", flowTx.ReferenceBlockID, err)
-			}
+	// TODO ensure the reference block is part of the main chain
+	_ = refBlock
```

Comment:
a bit confused by this instruction, can you explain why?

Reply:
Essentially to do this: https://github.com/onflow/flow-go/issues/4204.

Right now, headers from cluster vs main chain are not differentiated in the storage layer, so there is a potential attack vector where a reference block from a cluster chain is used here.